### Data engineering exploration: League of Legends edition

You might be wondering, where is the point of this project? The response is straightforward: I have been a huge fan of League of Legends since my high school days, so I thought that using data of this game could be great to add to my portfolio in data engineering. 

Despite there are so much data about this game scattered across the internet, I considered that it would be fun to create my own dataset with the most relevant information of my recent games using the Riot API, tool that allows us to collect a lot of data from every game of the company. 

We will need three libraries in this project:

* Requests: an elegant and simple HTTP library for Python that we will use to create the API calls.

* Pandas: one of the most famous Python libraries for data analysis and manipulation.

* Re: a library to work with regular expressions in Python.

Additionally, we import the API key alocated if 'creds' file, which enables us to make calls and extract data. Since it is linked to my account, it's crucial not to disclose it.

In [7]:
import requests
import pandas as pd
from creds import api_key
import re

### Part 1: SUMMONER-V4 Endpoint

In this part we will utilize  the SUMMONER-V4 endpoint to extract account data. My summoner name, or nickname, is 'Masgt7', so that is the account we will use in this demonstration.

In [8]:
url_summoner = 'https://euw1.api.riotgames.com/lol/summoner/v4/summoners/by-name/Masgt7' + '?api_key=' + api_key

r = requests.get (url_summoner)
r.status_code

200

We obtained 200 response code, which means that the call was succesfull. To see the data, we must use the json method

In [9]:
player_info = r.json()
player_info

{'id': '7xKp5xDyqk1u0_RKrh8SdaSHYhjpThiq7XXJscT2DQx273I',
 'accountId': 'NcGfPSobpg1gNOsZBjrQNCVyeIg7MyijVoglw0WBQioOST0',
 'puuid': 'uFIekhnHnZjaIXG_IG5y2Kb_UbBsEv57LksBUe6qMh0h0MPBKg8xoWtsbQkvXcPhlHLNGn8xZwqzGQ',
 'name': 'Masgt7',
 'profileIconId': 940,
 'revisionDate': 1701176501363,
 'summonerLevel': 75}

There is some information about my accout. In order to obtain the data about my recent games, we have to use the Platform Unique Identifier ('puuid') 

In [10]:
puuid = player_info['puuid']

### Part 2: MATCH-V5 Endpoint
Similar to the previous call, we will now extract the identifiers of my recent games. In the new URL for extracting this data we've included a new variable, 'num_games,' which represents the number of games to retrieve. It can range from 0 to 100, but for this demonstration we will select 20

In [11]:
num_games = 20

url_match_history = 'https://europe.api.riotgames.com/lol/match/v5/matches/by-puuid/' + puuid + '/ids?type=normal&start=0&count=' + str(num_games) + '&api_key=' + api_key
r = requests.get(url_match_history)

if r.status_code == 200:
    game_id = r.json()
    print (game_id)
else:
    print ('Error', r.status_code, ':',) 

['EUW1_6695825651', 'EUW1_6695809457', 'EUW1_6695778962', 'EUW1_6694669106', 'EUW1_6694631928', 'EUW1_6693543692', 'EUW1_6693492561', 'EUW1_6693435700', 'EUW1_6693383878', 'EUW1_6691848158', 'EUW1_6688283685', 'EUW1_6688256964', 'EUW1_6688236897', 'EUW1_6688218767', 'EUW1_6686773929', 'EUW1_6686744190', 'EUW1_6684157765', 'EUW1_6684133215', 'EUW1_6682813114', 'EUW1_6682767095']


With the identifiers of the games, we can now generate a list that contains dataframes created from the data of each match.

In [12]:
list_df = []

for index, identifier in enumerate(game_id):
    url_game_info = 'https://europe.api.riotgames.com/lol/match/v5/matches/' + identifier + '?api_key=' + api_key
    game_info = requests.get(url_game_info)
    index_player = game_info.json()['metadata']['participants'].index(puuid)
    game_data = game_info.json()['info']['participants'][index_player]
    game_df = pd.json_normalize(game_data)
    list_df.append(pd.DataFrame(game_df))

In [13]:
for i in range(len(game_id)):
    print (list_df[i].shape)

(1, 262)
(1, 264)
(1, 263)
(1, 263)
(1, 261)
(1, 263)
(1, 262)
(1, 263)
(1, 264)
(1, 264)
(1, 261)
(1, 263)
(1, 263)
(1, 263)
(1, 263)
(1, 262)
(1, 267)
(1, 268)
(1, 267)
(1, 268)


As we can see, this dataframes don't have the same columns, so we can't consolidate them in one. To achieve this, we will drop any columns that do not appear in every dataframe

In [14]:
common_columns = list(set.intersection(*(set(df.columns) for df in list_df)))

game_history = pd.concat(list_df)
game_history = game_history [common_columns]

game_history.shape

(20, 260)

The result is a dataframe with 260 columns. Let's take a look at the first five rows in the next cell.

In [15]:
game_history.head()

challenges.outnumberedNexusKill  challenges.takedownOnFirstTurret  \
0                                0                                 1   
0                                0                                 1   
0                                0                                 1   
0                                0                                 0   
0                                0                                 1   

   challenges.damagePerMinute  largestMultiKill  inhibitorKills  \
0                  966.848894                 2               0   
0                 1197.243384                 1               0   
0                 1023.219236                 1               0   
0                 1040.204218                 2               0   
0                  946.131503                 2               0   

   challenges.goldPerMinute  firstTowerAssist  longestTimeSpentLiving  \
0                672.075885             False                     151   
0                644.579756             False                     226   
0                713.007854             False                     194   
0                762.588655              True                     244   
0                664.695674             False                     234   

   challenges.quickFirstTurret  challenges.baronTakedowns  ...  \
0                            0                          0  ...   
0                            0                          0  ...   
0                            0                          0  ...   
0                            1                          0  ...   
0                            0                          0  ...   

   challenges.firstTurretKilled  challenges.buffsStolen  visionScore  \
0                             0                       0            0   
0                             1                       0            0   
0                             1                       0            0   
0                             1                       0            0   
0                             0                       0            0   

   challenges.deathsByEnemyChamps  firstBloodKill  \
0                               9           False   
0                               8           False   
0                               9           False   
0                               5           False   
0                              13           False   

   challenges.dodgeSkillShotsSmallWindow  playerScore2  itemsPurchased  \
0                                      0             0              20   
0                                      0             0              12   
0                                      0             0              20   
0                                      0             0              15   
0                                      6             0              20   

   totalDamageDealt  enemyMissingPings  
0             39560                  0  
0             47442                  0  
0            116929                  0  
0             35165                  0  
0             96251                  1  

[5 rows x 260 columns]

For the sake of clarity, we will remove the text preceding the dot in columns where it occurs . We will do this using regular expressions

In [16]:
def remove_dot (string):
    pattern = r'.*?\.(.*)'
    result = re.match(pattern, string)
    if result:
        return result.group(1)
    else:
        return string

In [17]:
df_columns = list(game_history.columns)

for i,j in enumerate(df_columns):
    df_columns[i] = remove_dot(j)

game_history.columns = df_columns

game_history.head(1)

outnumberedNexusKill  takedownOnFirstTurret  damagePerMinute  \
0                     0                      1       966.848894   

   largestMultiKill  inhibitorKills  goldPerMinute  firstTowerAssist  \
0                 2               0     672.075885             False   

   longestTimeSpentLiving  quickFirstTurret  baronTakedowns  ...  \
0                     151                 0               0  ...   

   firstTurretKilled  buffsStolen  visionScore  deathsByEnemyChamps  \
0                  0            0            0                    9   

   firstBloodKill  dodgeSkillShotsSmallWindow  playerScore2  itemsPurchased  \
0           False                           0             0              20   

   totalDamageDealt  enemyMissingPings  
0             39560                  0  

[1 rows x 260 columns]

The API provides detailed information about each game, but for clarity, we will select only a subset of columns with the most relevant ones

In [18]:
game_history = game_history [['kills', 'deaths', 'assists', 'kda', 'championName', 'individualPosition', 'killParticipation', 'summoner1Casts', 'summoner2Casts', 
                       'largestMultiKill', 'totalDamageDealtToChampions', 'teamDamagePercentage', 'totalDamageTaken', 'damageTakenOnTeamPercentage',
                       'largestKillingSpree', 'allInPings', 'assistMePings', 'baitPings', 
                       'basicPings', 'commandPings', 'dangerPings', 'enemyVisionPings', 'getBackPings', 'visionClearedPings', 'holdPings', 'needVisionPings', 
                       'onMyWayPings', 'gameLength', 'goldEarned', 'goldPerMinute', 'wardsPlaced', 'wardsKilled', 'win']]

In [19]:
game_history.shape

(20, 33)

The result of this selection is that we have passed from an enormous dataframe with high dimensionality to a smaller one with 33 columns that describe my performance in the games in a concise way

Now, we will change the index of each row in order to see directly the game that it describes.

In [20]:
game_history.reset_index(drop=True, inplace=True)
game_history.index += 1

game_history.head()

kills  deaths  assists       kda championName individualPosition  \
1      5       9       29  3.777778       KSante            Invalid   
2      5       8       22  3.375000      Leblanc            Invalid   
3      8       9       27  3.888889      Caitlyn            Invalid   
4      8       5       26  6.800000       KSante            Invalid   
5      8      13       38  3.538462      Morgana            Invalid   

   killParticipation  summoner1Casts  summoner2Casts  largestMultiKill  ...  \
1           0.918919               4              18                 2  ...   
2           0.642857               1              13                 1  ...   
3           0.583333               4               3                 1  ...   
4           0.641509               3              17                 2  ...   
5           0.718750               5              15                 2  ...   

   visionClearedPings  holdPings  needVisionPings  onMyWayPings   gameLength  \
1                   0          0                0             0   969.932236   
2                   0          0                0             0  1086.049539   
3                   0          0                0             0  1413.845533   
4                   0          0                0             1   900.228990   
5                   0          0                0             2  1543.271572   

   goldEarned  goldPerMinute  wardsPlaced  wardsKilled    win  
1       10864     672.075885            0            0  False  
2       11667     644.579756            0            0  False  
3       16801     713.007854            0            0  False  
4       11441     762.588655            0            0   True  
5       17096     664.695674            0            0  False  

[5 rows x 33 columns]

Finally, we will do some transformations in some variables to provide a clearer view of them.

We multiply the percentages by 100 and round them

In [26]:
percentages = ['killParticipation', 'teamDamagePercentage', 'damageTakenOnTeamPercentage']

for i in percentages:
        game_history[i] = round (game_history[i] * 100, 2);

Also, aiming to reduce even more the dimensionality we can sum all the pings in just one column. Pings in LoL are important because they are an efficient way to comunicate your team several events in the game, but in our example we do not need data of them with this detail

In [22]:
pings = game_history.loc[:, 'allInPings':'onMyWayPings']
totalPings = pings.sum(axis=1)

game_history.insert(15, 'totalPings', totalPings)

game_history = game_history.drop(pings, axis=1)

Lastly, we will transform the column containing the duration of the game from seconds to minutes and round two columns with float values

In [23]:
game_history['gameLength'] = pd.to_datetime(game_history['gameLength'], unit = 'm').dt.strftime('%H:%M')

game_history['goldPerMinute'] = round (game_history['goldPerMinute'], 2)

game_history['kda'] = round (game_history['kda'], 2)

Finally, here is the resulting dataset that we saved as a CSV file in the project root.

In [24]:
pd.set_option('display.max_columns', None)
display(game_history)

kills  deaths  assists    kda championName individualPosition  \
1       5       9       29   3.78       KSante            Invalid   
2       5       8       22   3.38      Leblanc            Invalid   
3       8       9       27   3.89      Caitlyn            Invalid   
4       8       5       26   6.80       KSante            Invalid   
5       8      13       38   3.54      Morgana            Invalid   
6      12       9       22   3.78       LeeSin            Invalid   
7      11      11       26   3.36        Xayah            Invalid   
8       8      15       36   2.93      Morgana            Invalid   
9      12       6       30   7.00        Corki            Invalid   
10     23       5       16   7.80     Tristana            Invalid   
11      6      10       28   3.40         Shen            Invalid   
12     15      13       24   3.00        Xayah            Invalid   
13     19      10        9   2.80          Jax            Invalid   
14      6      10       14   2.00      Camille            Invalid   
15      8       6       28   6.00        Corki            Invalid   
16     12       8       10   2.75      Caitlyn            Invalid   
17     14       8       16   3.75      Camille            Invalid   
18     17       2       17  17.00         Jhin            Invalid   
19      6      10       45   5.10        Rakan            Invalid   
20     22      11       26   4.36         Ahri            Invalid   

    killParticipation  summoner1Casts  summoner2Casts  largestMultiKill  \
1            0.918919               4              18                 2   
2            0.642857               1              13                 1   
3            0.583333               4               3                 1   
4            0.641509               3              17                 2   
5            0.718750               5              15                 2   
6            0.618182               3              15                 2   
7            0.569231               4               3                 1   
8            0.687500               4              17                 2   
9            0.857143               5               8                 3   
10           0.750000               2               3                 4   
11           0.772727               5              33                 1   
12           0.609375               5               6                 2   
13           0.549020               6              36                 2   
14           0.500000               4              16                 2   
15           0.720000               4               5                 2   
16           0.488889               3               3                 2   
17           0.508475               4              18                 4   
18           0.755556               3               2                 4   
19           0.784615               3              23                 2   
20           0.738462               6              20                 3   

    totalDamageDealtToChampions  teamDamagePercentage  totalDamageTaken  \
1                         15629                 16.42             28142   
2                         21671                 19.19             16912   
3                         24111                 16.06             22086   
4                         15607                 14.08             22333   
5                         24335                 13.06             31926   
6                         20499                 20.19             17429   
7                         45642                 22.92             31887   
8                         27141                 14.50             39686   
9                         60120                 30.14             20797   
10                        23585                 22.87             12940   
11                        25707                 15.52             58100   
12                        27617                 22.45             30754   
13 

In [25]:
game_history.to_csv('game_history.csv', mode = 'x', sep=',')

That's all. Thank you for reading and checking my project!